### PR TITLE
[WIP][tune] Enable training function FT and Trainable checkpoint policies

### LINKED ISSUE #6162: [tune] Support flexible checkpointing in training functions and Trainables
Support some subset of: 
1. ~checkpoints to be taken mid-`_train` call~
2. checkpointing/restoring from within training functions (rather than just in `Trainable`)
3. more flexible checkpoint policies (rather than just passing in frequency). 

**Proposal** 

Function interface
```python
def train(config):
  if tune.track.can_load():
    model = restore_my_model(tune.track.load())  # restore from dict or path
  else:
    model = init_my_model()

  for i in range(start_iter, 100):
    ...
    result = {"acc": acc}
    if i % 10 == 0:
      # (1) save to files in dir
      checkpoint_dir = tune.track.get_checkpoint_dir()
      save_my_model(model, checkpoint_dir)
      # (2) OR save dict
      checkpoint = {"state": model.state()}
    
    tune.track.log(result, save=checkpoint)

tune.run(train)
```

Trainable interface
```python
class CheckpointPolicy:

  def should_checkpoint(result):
    """Returns whether or not checkpoint should be taken.
    Result is none for first call (before training begins).
    """
    raise NotImplementedError

  def checkpoint_score(result):
    """Returns the score of the checkpoint corresponding to the result."""
    raise NotImplementedError

tune.run(my_trainable, checkpoint_policy=PeriodicCheckpointPolicy(frequency=5))    
```

### PR DESCRIPTION
**Pull request by ujvl**
## Why are these changes needed?

**Functions**

Support checkpoint/restore functionality in training functions so that users that don't want to use the `Trainable` interface can still benefit from FT. 

**Trainables**

Also make it easier to implement more expressive checkpointing policies (as opposed to just specifying a frequency) for consistent functionality.

See interface broadly defined in #6162.

TODO
- [ ] figure out how to guarantee checkpoint integrity
- [ ] finish checkpoint/restore for functions
- [ ] add tests
- [ ] test manually (local and on cluster)

## Related issue number

Closes #6162 

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: python/ray/tune/checkpoint_policy.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+        Args:
+            result (dict): result from last training step.
+
+        Returns:
+            True if should take checkpoint, False otherwise.
+        """
+        raise NotImplementedError
+
+    def checkpoint_score(self, result):
```
Comment (hartikainen) [inline python/ray/tune/checkpoint_policy.py]: Should this be more general on this level, i.e. `def garbage_collect` or something like that?

File: python/ray/tune/tune.py [hunk 2]
```diff
@@ -135,10 +136,13 @@ def run(run_or_experiment,
             remote node to local. If string, then it must be a string template
             that includes `{source}` and `{target}` for the syncer to run.
             If not provided, defaults to using rsync.
-        checkpoint_freq (int): How many training iterations between
-            checkpoints. A value of 0 (default) disables checkpointing.
-        checkpoint_at_end (bool): Whether to checkpoint at the end of the
-            experiment regardless of the checkpoint_freq. Default is False.
+        checkpoint_policy (CheckpointPolicy): yolo.
```
Comment (hartikainen) [inline python/ray/tune/tune.py]: yolo

File: python/ray/tune/checkpoint_policy.py (hunk 1 above)
Comment (ujvl) [reply python/ray/tune/checkpoint_policy.py]: yeah i need to refactor this, been busy with other stuff

File: python/ray/tune/tune.py (hunk 2 above)
Comment (ujvl) [reply python/ray/tune/tune.py]: whoops WIP

**Comment by richardliaw** [comment]
Excited to see this merged, @ujvl!

**Comment by ujvl** [comment]
Closed in favor of #7113

**Pull request closed without merging**